File: src/main/javascript/@node/nest/keeper.ts
/*
 * Copyright © 2020-2022 Metreeca srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Optional } from "@metreeca/core";
import { Context, createContext, createElement, ReactNode, useContext } from "react";


/**
 * Keeper context.
 *
 * Provides nested components with a {@link useKeeper| head}-based shared state containing:
 *
 * - the current user profile
 *
 * @module
 */

const Context=createContext<[Value, Updater]>([undefined, () => {}]);


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
 * The value component of the keeper context state.
 *
 * Holds a user profile of type `P`, if a user is authenticated; `undefined`, otherwise
 *
 *  @template P the type of the user profile
 */
export type Value<P=unknown>=Optional<P>;

/**
 * The updater component of the keeper context state.
 *
 *  @template P the type of the user profile
 */
export interface Updater<P=unknown> {

    (): void;

    (anonymous: null): void;

    (authenticated: P): void;

}


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
 * Creates a keeper context.
 *
 * @param children the children components to be nested inside the new context component
 *
 * @return a new keeper context component
 *
 */
export function NodeKeeper<P>({

    state,

    children

}: {

    state: [Value<P>, Updater<P>]

    children: ReactNode

}) {

    return createElement((Context as Context<[Value<P>, Updater<P>]>).Provider, {

        value: state,

        children

    });

}

/**
 * Creates a {@link NodeKeeper| keeper context} hook.
 *
 * @return a state tuple including a stateful {@link Value| value} and an {@link Updater| updater} function.
 *
 * @template P the type of the user profile
 */
export function useProfile<P>(): [Value<P>, Updater<P>] {
    return useContext<[Value<P>, Updater<P>]>(Context as any);
}

File: src/main/java/eu/ec2u/card/filters/Keeper.java
/*
 * Copyright © 2020-2022 EC2U Alliance
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package eu.ec2u.card.filters;

import com.google.inject.Inject;
import com.sun.net.httpserver.Filter;
import com.sun.net.httpserver.HttpExchange;
import eu.ec2u.card.Setup;
import eu.ec2u.card.handlers.Root.Holder;
import eu.ec2u.card.services.Notary;
import lombok.experimental.Delegate;

import java.io.IOException;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static eu.ec2u.card.Handler.Unauthorized;
import static eu.ec2u.card.Handler.holder;

import static java.lang.String.format;

public final class Keeper extends Filter {

    private static final Pattern BearerPattern=Pattern.compile("\\s*Bearer\\s*(?<token>\\S*)\\s*");


    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Inject private Setup setup;
    @Inject private Notary notary;


    @Override public String description() {
        return "bearer token authenticator";
    }

    @Override public void doFilter(final HttpExchange exchange, final Chain chain) throws IOException {

        Optional.ofNullable(exchange.getRequestHeaders().getFirst("Authorization"))

                .map(BearerPattern::matcher)
                .filter(Matcher::matches)
                .map(matcher -> matcher.group("token"))

                .map(this::validate) // !!! malformed/expired

                .ifPresent(holder -> holder(exchange, holder));

        chain.doFilter(new ExchangeWrapper(exchange) {

            @Override public void sendResponseHeaders(final int rCode, final long responseLength) throws IOException {

                if ( rCode == Unauthorized ) {
                    getResponseHeaders().set("WWW-Authenticate", format("Bearer realm=\"%s\"", setup.getSso()));
                }

                super.sendResponseHeaders(rCode, responseLength);
            }

        });

    }


    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private Holder validate(final String token) {
        return new Holder()
                .setEsi("urn:schac:personalUniqueCode:int:esi:unipv.it:999001")
                .setUni("unipv.it");
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private abstract static class ExchangeWrapper extends HttpExchange {

        @Delegate private final HttpExchange exchange;

        ExchangeWrapper(final HttpExchange exchange) { this.exchange=exchange; }

    }

}
